### PR TITLE
 python310Packages.coincurve: 17.0.0 -> 18.0.0 

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by prusnak** [comment]
NACK. We should fix Electrum first.

**Comment by gruve-p** [comment]
NACK
This breaks Electrum-GRS, we need to address this before this PR gets merged.

**Comment by lourkeur** [comment]
Regarding Electrum and variants, it seems like the problem is that the secp256k1 derivation now calls its library `libsecp256k1.so.1.0.0` instead of `libsecp256k1.so.0`, which is what Electrum expects. A non-Nix installation of Electrum with libsecp256k1 0.2 does not have this problem, so it looks like a bug on our side.

**Comment by SuperSandro2000** [comment]
Are there any open PRs I can review and merge to fix this?

**Comment by prusnak** [comment]
> Are there any open PRs I can review and merge to fix this?

I don't think so.

> Otherwise I would mark them broken due to lack of time on my side.

If you don't have time for this, we should not update the package that does the breakage.

**Comment by prusnak** [comment]
Above said, I will have a look at the issue tomorrow and fix Electrum. The bug is inside Nixpkgs after updating secp256k1, not in the Electrum upstream.

**Comment by lourkeur** [comment]
I dug a bit deeper. It looks like https://github.com/bitcoin-core/secp256k1/commit/e025ccdf7473702a76bb13d763dc096548ffefba changed the ABI version on purpose. It's not a bug on our side.

libbitcoin has its own fork of [libsecp256k1](https://github.com/libbitcoin/secp256k1) so their package could use that. Or we could patch it.

Regarding Electrum and derivatives, they're missing the `recover` module.

File: pkgs/tools/security/secp256k1/default.nix [hunk 1]
```diff
@@ -4,27 +4,22 @@
...
 
   nativeBuildInputs = [ autoreconfHook ];
 
   configureFlags = [
     "--enable-benchmark=no"
     "--enable-exhaustive-tests=no"
-    "--enable-experimental"
-    "--enable-module-ecdh"
-    "--enable-module-recovery"
-    "--enable-module-schnorrsig"
```
Comment (lourkeur) [inline pkgs/tools/security/secp256k1/default.nix:27]: For best results with Electra, don't disable the recovery module. Is there a reason to remove the other 2?

File: pkgs/applications/misc/electrum/default.nix [hunk 2]
```diff
@@ -154,5 +132,7 @@ python3.pkgs.buildPythonApplication {
     license = licenses.mit;
     platforms = platforms.all;
     maintainers = with maintainers; [ joachifm np prusnak ];
+    # requires libsecp256k1 < 0.2.0
+    broken = true;
```
Comment (lourkeur) [inline pkgs/applications/misc/electrum/default.nix:136]:
```suggestion
```

File: pkgs/applications/misc/electrum/ltc.nix [hunk 3]
```diff
@@ -158,5 +136,7 @@ python3.pkgs.buildPythonApplication {
     license = licenses.mit;
     platforms = platforms.all;
     maintainers = with maintainers; [ lourkeur ];
+    # requires libsecp256k1 < 0.2.0
+    broken = true;
```
Comment (lourkeur) [inline pkgs/applications/misc/electrum/ltc.nix:140]:
```suggestion
```

File: pkgs/applications/misc/electrum-grs/default.nix [hunk 4]
```diff
@@ -102,5 +102,7 @@ python3.pkgs.buildPythonApplication {
     license = licenses.mit;
     platforms = platforms.all;
     maintainers = with maintainers; [ gruve-p ];
+    # requires libsecp256k1 < 0.2.0
```
Comment (lourkeur) [inline pkgs/applications/misc/electrum-grs/default.nix:105]:
```suggestion
    # requires protobuf < 4
```
Hit another issue

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/electrum-grs/default.nix (hunk 4 above)
Comment (gruve-p) [reply pkgs/applications/misc/electrum-grs/default.nix:105]: Fix in #207752

**Comment by lourkeur** [comment]
@SuperSandro2000 can you rebase? Otherwise I can take over.

File: pkgs/tools/security/secp256k1/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/tools/security/secp256k1/default.nix:27]: ecdh and schnorrsig are on by default. I don't think we want experimental on unless required. Going to add recoverry back.

File: pkgs/tools/security/secp256k1/default.nix [hunk 5]
```diff
@@ -4,27 +4,22 @@
...
-    sha256 = "x9qG2S6tBSRseWaFIN9N2fRpY1vkv8idT3d3rfJnmaU=";
+    rev = "v${version}";
+    sha256 = "sha256-wYJIMCoo6ryeQN4ZnvEkJ5/332+AkaOwgplDuQQC5MU=";
   };
 
   nativeBuildInputs = [ autoreconfHook ];
 
   configureFlags = [
     "--enable-benchmark=no"
     "--enable-exhaustive-tests=no"
```
Comment (SuperSandro2000) [inline pkgs/tools/security/secp256k1/default.nix:22]:
```suggestion
    "--enable-exhaustive-tests=no"
    "--enable-module-recovery"
```

**Comment by SuperSandro2000** [comment]
rebased and cleaned up

File: pkgs/applications/misc/electrum/ltc.nix [hunk 6]
```diff
@@ -134,5 +134,7 @@ python3.pkgs.buildPythonApplication {
...
```
Comment (lourkeur) [inline pkgs/applications/misc/electrum/ltc.nix:138]:
```suggestion
```

builds and runs just as well as the other 2

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 206150` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>electrum-ltc</li>
    <li>python310Packages.bip_utils</li>
    <li>python311Packages.bip_utils</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>echidna</li>
    <li>haskellPackages.libsecp256k1</li>
    <li>python311Packages.cairo-lang</li>
    <li>python311Packages.eth-account</li>
    <li>python311Packages.eth-keyfile</li>
    <li>python311Packages.eth-keys</li>
    <li>python311Packages.web3</li>
  </ul>
</details>
<details>
  <summary>26 packages built:</summary>
  <ul>
    <li>electron-cash</li>
    <li>electrum</li>
    <li>electrum-grs</li>
    <li>haskellPackages.bitcoin-scripting</li>
    <li>haskellPackages.haskoin-core</li>
    <li>haskellPackages.haskoin-node</li>
    <li>haskellPackages.haskoin-store-data</li>
    <li>haskellPackages.hevm</li>
    <li>haskellPackages.secp256k1-haskell</li>
    <li>jami-client</li>
    <li>jami-daemon</li>
    <li>libbitcoin</li>
    <li>libbitcoin-client</li>
    <li>libbitcoin-explorer</li>
    <li>libbitcoin-network</li>
    <li>libbitcoin-protocol</li>
    <li>python310Packages.cairo-lang</li>
    <li>python310Packages.coincurve</li>
    <li>python310Packages.eth-account</li>
    <li>python310Packages.eth-keyfile</li>
    <li>python310Packages.eth-keys</li>
    <li>python310Packages.secp256k1</li>
    <li>python310Packages.web3</li>
    <li>python311Packages.coincurve</li>
    <li>python311Packages.secp256k1</li>
    <li>secp256k1</li>
  </ul>
</details>

Not sure if the failures are related

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
Python 3.11 Eth-hash will fix itself from upstream so this is not a concern for this PR https://github.com/ethereum/eth-hash/commit/176089f841a8d167f1829f7fd3aed1eb9d9a9a2f

The Haskell `libsecp256k1` failure appears to be probabilistic.

---

Result of `nixpkgs-review pr 206150` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>electrum-ltc</li>
    <li>python310Packages.bip_utils</li>
    <li>python311Packages.bip_utils</li>
  </ul>
</details>
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>python311Packages.cairo-lang</li>
    <li>python311Packages.eth-account</li>
    <li>python311Packages.eth-keyfile</li>
    <li>python311Packages.eth-keys</li>
    <li>python311Packages.web3</li>
  </ul>
</details>
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>echidna</li>
    <li>electron-cash</li>
    <li>electrum</li>
    <li>electrum-grs</li>
    <li>haskellPackages.bitcoin-scripting</li>
    <li>haskellPackages.haskoin-core</li>
    <li>haskellPackages.haskoin-node</li>
    <li>haskellPackages.haskoin-store-data</li>
    <li>haskellPackages.hevm</li>
    <li>haskellPackages.libsecp256k1</li>
    <li>haskellPackages.secp256k1-haskell</li>
    <li>jami-client</li>
    <li>jami-daemon</li>
    <li>libbitcoin</li>
    <li>libbitcoin-client</li>
    <li>libbitcoin-explorer</li>
    <li>libbitcoin-network</li>
    <li>libbitcoin-protocol</li>
    <li>python310Packages.cairo-lang</li>
    <li>python310Packages.coincurve</li>
    <li>python310Packages.eth-account</li>
    <li>python310Packages.eth-keyfile</li>
    <li>python310Packages.eth-keys</li>
    <li>python310Packages.secp256k1</li>
    <li>python310Packages.web3</li>
    <li>python311Packages.coincurve</li>
    <li>python311Packages.secp256k1</li>
    <li>secp256k1</li>
  </ul>
</details>

**Comment by lourkeur** [comment]
Superseded by #208871 and #208232

**Comment by SuperSandro2000** [comment]
and #214212 and #214213

**Pull request closed without merging**